### ISSUE TITLE
Commands can stop working if no foreground window

### ISSUE THREAD

**Issue by vegardlarsen** [digitalcreations/MaxTo#394]
**Describe the bug**
MaxTo commands can crash if [GetForegroundWindow](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-getforegroundwindow) returns `null`.

**To Reproduce**
Not consistent, but executing e.g. "Move window right" in quick succession seems to trigger it at least on one customer's machine.

**Expected behavior**
MaxTo should fall back to using the window z-index to find the topmost window if there is no foreground window.

**System information:**
 - Windows version: 10
 - MaxTo version 2.0.1

**Additional context**
```text
2019-04-11 07:07:43 [Core@2.0.1.499] [Information] Executing recipe '"Move window down"' with 1 ingredients
2019-04-11 07:07:43 [Core@2.0.1.499] [Information] Recipe '"Move window down"' executed successfully.
2019-04-11 07:07:43 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:07:43 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:07:44 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:07:44 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 07:07:45 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:07:45 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 07:07:48 [UserInterface@2.0.1.499] [Information] Found 2 screens.
2019-04-11 07:08:00 [UserInterface@2.0.1.499] [Information] Updated regions on "TVWSJD\Generic PnP Monitor\{BBD3BEE4-3AD5-11E9-8147-484D7EE6E4E6}\0000\3840@96x2160@96\95756c69-c075-4f90-9e9e-f1692489050a"
2019-04-11 07:08:01 [UserInterface@2.0.1.499] [Information] Updated regions on "TVWSJD\Generic PnP Monitor\{BBD3BEE4-3AD5-11E9-8147-484D7EE6E4E6}\0001\3840@96x2160@96\95756c69-c075-4f90-9e9e-f1692489050a"
2019-04-11 07:08:04 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:08:05 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:08:07 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:08:07 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 07:08:07 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:08:07 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:08:10 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:08:10 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:08:10 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:08:10 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:08:11 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:08:12 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:08:15 [Core@2.0.1.499] [Information] Executing recipe '"Move window down"' with 1 ingredients
2019-04-11 07:08:15 [Core@2.0.1.499] [Information] Recipe '"Move window down"' executed successfully.
2019-04-11 07:08:15 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:08:15 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:08:16 [Core@2.0.1.499] [Information] Executing recipe '"Move window down"' with 1 ingredients
2019-04-11 07:08:16 [Core@2.0.1.499] [Information] Recipe '"Move window down"' executed successfully.
2019-04-11 07:08:34 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:08:34 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 07:08:34 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:08:34 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:12:32 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:12:32 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Executing recipe '"Move window down"' with 1 ingredients
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Recipe '"Move window down"' executed successfully.
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:12:33 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:13:03 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:13:03 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:13:04 [Core@2.0.1.499] [Information] Executing recipe '"Move window up"' with 1 ingredients
2019-04-11 07:13:04 [Core@2.0.1.499] [Information] Recipe '"Move window up"' executed successfully.
2019-04-11 07:26:38 [Core@2.0.1.499] [Information] Executing recipe '"Move window down"' with 1 ingredients
2019-04-11 07:26:38 [Core@2.0.1.499] [Information] Recipe '"Move window down"' executed successfully.
2019-04-11 07:26:39 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 07:26:39 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 07:26:39 [Core@2.0.1.499] [Information] Executing recipe '"Move window left"' with 1 ingredients
2019-04-11 07:26:39 [Core@2.0.1.499] [Information] Recipe '"Move window left"' executed successfully.
2019-04-11 08:21:18 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 08:21:18 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 08:21:18 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 08:21:18 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 08:21:20 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 08:21:20 [Core@2.0.1.499] [Information] Recipe '"Move window right"' executed successfully.
2019-04-11 10:29:08 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 10:29:08 [Core@2.0.1.499] [Warning] Received exception from command request 1 while executing '"Move window right"'. Aborting.
System.ArgumentException: Could not find monitor matching 
   at MaxTo.Core.Commands.Window.MoveCommand.<Execute>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MaxTo.Core.Commands.CommandExecutor.<Execute>d__3`2.MoveNext()
2019-04-11 10:29:08 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 10:29:08 [Core@2.0.1.499] [Warning] Received exception from command request 1 while executing '"Move window right"'. Aborting.
System.ArgumentException: Could not find monitor matching 
   at MaxTo.Core.Commands.Window.MoveCommand.<Execute>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MaxTo.Core.Commands.CommandExecutor.<Execute>d__3`2.MoveNext()
2019-04-11 10:29:08 [Core@2.0.1.499] [Information] Executing recipe '"Move window right"' with 1 ingredients
2019-04-11 10:29:08 [Core@2.0.1.499] [Warning] Received exception from command request 1 while executing '"Move window right"'. Aborting.
System.ArgumentException: Could not find monitor matching 
   at MaxTo.Core.Commands.Window.MoveCommand.<Execute>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MaxTo.Core.Commands.CommandExecutor.<Execute>d__3`2.MoveNext()
```

**Comment by boniboni** [comment]
I can confirm this happens in my machine as well

**Comment by vegardlarsen** [comment]
We have not been able to track this down further. I was a bit naive earlier; it is not possible to use the z-order to find a foreground window if there currently is none. When Windows reports that there isn't a foreground window, they are right. :)

We have not been able to find a root cause for this either. For now, we've added better logging around the area that produces this exception, so we can hopefully track this down in the future.

**Comment by vegardlarsen** [comment]
Closing this issue as it does not appear to happen often, and may already have been fixed in the 2.1.0 previews. Please correct me if that assumption is wrong.
